### ISSUE TITLE
packet: Consider using afterburn to simplify setup_raid_* flags

### ISSUE THREAD

**Issue by rata** [kinvolk/lokomotive-kubernetes#75]
Right now we are using a bash script in the ignition configuration to create RAID arrays in Packet. This script is getting messier and doesn't feel like a clean way to implement this:  https://github.com/kinvolk/lokomotive-kubernetes/blob/00e6d550c369035b4f0b57c7ab46d01875207c9d/packet/flatcar-linux/kubernetes/workers/cl/worker.yaml.tmpl#L128-L223

You can see it's ~100 LOC, created several bugs (just see the history in that file) and before it gets out of control we should consider options to change this.

On other platforms we can probably just use ignition as the disk layout is stable. In packet, however, this is not the case: /dev/sda might be the smallest of 4 disks and rotational on one server while in another identical server /dev/sda might be the biggest of disks and even ssd.  Therefore, we couldn't use ignition as it lacks the expresiveness to selec disks with those criterias (smallest, ssd, hdd).

However, I think we might be able to simplify all of this and use ignition if we add more features to afterburn for the Packet provider.

I'm not familiar with [afterburn](https://github.com/flatcar-linux/afterburn) but it seems like a nice way to export metadata and **maybe** (not sure it really makes sense) we can consider exporting some other variables in packet, like:

 * AFTERBURN_PACKET_DISKS_SMALLEST: this variable can be the smallest disk on that instance, detected by afterburn, or all the smallest disks and the user can select one of them

* AFTERBURN_PACKET_DISKS_SSD: the list of SSD disks in the server that are not in use (i.e. excluding the disk used for the OS or that are used for something else), for example "/dev/sda /dev/sdd /dev/sdb"

* AFTERBURN_PACKET_DISKS_HDD`: the list of HDD disks in the server that are not in use (i.e. excluding the disk used for the OS or that are used for something else), for example "/dev/sda /dev/sdd /dev/sdb"

**If this is possible and these variables can be used in ignition to format the disks**, we should be able to dramatically simplify the `setup_raid_*` flags we currently have.

Please note that this will not solve https://github.com/kinvolk/lokomotive-kubernetes/issues/73, this will just remove this every time more complicated bash script before it runs out of control, probably reduce quite some bugs there and, hopefully, implement this in a clean way.

**This can also probably remove from flatcar-install the `-s` experimental functionality** created here https://github.com/flatcar-linux/init/commit/1a2453218efc10e5437cde9e7a9dd0eae156d770, as now flatcar install will be possible to be called (if afterburn is enabled) with something like:

```
flatcar-install -d ${AFTERBURN_PACKET_DISKS_SMALLEST}
```
And, therefore, the `-s` functionality for packet can be removed (the only platform we are using it right now).

**Comment by dongsupark** [comment]
I think in general it's a good idea.
Though in practice, it needs some extra work to do that.

First, each attr name should have a `COREOS_` prefix, following the [`cl-legacy` naming scheme](https://github.com/flatcar-linux/afterburn/blob/flatcar-master/docs/container-linux-legacy.md). The `AFTERBURN_` prefix is being used only for Fedora CoreOS.

Ideally we should probably rename every `COREOS_` to `FLATCAR_` from the afterburn repo. I'm not sure if that will require changes in other places as well.

**Comment by rata** [comment]
@dongsupark oh, right! We are in fact using `COREOS_` ones. I think we should keep using `COREOS_` to really be a drop-in replacement, though, but that is another story.

The idea of this issue is to see if it makes sense to have something like that in afterburn, we can decide on vars name later. But thanks a lot for noticing there is possible some extra work to do  =)

**Comment by johananl** [comment]
Some thoughts on the matter:

- Can we run Afterburn before Ignition runs? If not, I guess the "disk classification" logic should reside in Ignition itself.
- If we implement such a logic, ideally it should support arbitrary disk configurations. For example, it should be possible to perform the following disk "selections":
  - Smallest disk
  - Largest disk
  - Two smallest disks (assuming there are multiple disks of the smallest size, use the first two of them)
  - All smallest disks (assuming there are multiple disks of the smallest size, use all of them)
  - All disks of a certain type (e.g. SSD)
  - All disks
- The previous point makes me think it may be too complicated to include all possible selectors as Afterburn variables, so maybe it makes sense to have this entire logic in Ignition and not in Afterburn.
- On my testing, the Packet metadata service doesn't show all disks. It may be safer and more portable to check the actual disks on the machine.

**Comment by rata** [comment]
> Some thoughts on the matter:
> 
>     * Can we run Afterburn before Ignition runs? If not, I guess the "disk classification" logic should reside in Ignition itself.

Yeah, I'm not sure. That is what I meant with: **If this is possible and these variables can be used in ignition to format the disks**.

Sorry I wasn't clear.

>     * If we implement such a logic, ideally it should support arbitrary disk configurations. For example, it should be possible to perform the following disk "selections":
>       
>       * Smallest disk
>       * Largest disk
>       * Two smallest disks (assuming there are multiple disks of the smallest size, use the first two of them)
>       * All smallest disks (assuming there are multiple disks of the smallest size, use all of them)
>       * All disks of a certain type (e.g. SSD)
>       * All disks
> 
>     * The previous point makes me think it may be too complicated to include all possible selectors as Afterburn variables, so maybe it makes sense to have this entire logic in Ignition and not in Afterburn.

I think there is syntax available to express that on any of those, so that would be tricky.

> 
>     * On my testing, the Packet metadata service doesn't show all disks. It may be safer and more portable to check the actual disks on the machine.

Yeah, my (hacky) think was that the metadata service, when running on packet, will check that in the current server (instead of using packet metadata server, just poking the server instance via go) and expose the variable.

It's not metadata exposed by packet metadata service, but is still metadata we need. That was my thinking.

Regarding adding it to ignition, please note that Ignition is still supported by upstream and, IIUC, we prefer to keep doing that. So, we will need to send discuss such a change with upstream.

**Comment by johananl** [comment]
>Yeah, my (hacky) think was that the metadata service, when running on packet, will check that in the current server (instead of using packet metadata server, just poking the server instance via go) and expose the variable.

This makes sense to me only if the disk-related metadata is needed by user applications. AFAICT it will only be needed by Ignition, in which case it could make more sense to have the logic in Ignition and not Afterburn.
